schema/validator.go

```diff
@@ -73,3 +86,35 @@ type unimplemented string
 func (v unimplemented) Validate(src io.Reader) error {
 	return fmt.Errorf("%s: unimplemented", v)
 }
+
+func validateManifestDescendants(r io.Reader) error {
+	header := v1.Manifest{}
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors.Wrapf(err, "error reading the io stream")
+	}
+
+	err = json.Unmarshal(buf, &header)
+	if err != nil {
+		return errors.Wrap(err, "manifest format mismatch")
+	}
+
+	if header.Config.MediaType != string(v1.MediaTypeImageConfig) &&
+		header.Config.MediaType != MediaTypeDockerImageConfig {
+		return fmt.Errorf("illegal config mediaType: %s", header.Config.MediaType)
+	}
+
+	for _, layer := range header.Layers {
+		if layer.MediaType != string(v1.MediaTypeImageLayer) &&
+			layer.MediaType != MediaTypeDockerImageLayer &&
+			layer.MediaType != MediaTypeNonDistributable {
+			return fmt.Errorf("illegal layer mediaType: %s", layer.MediaType)
+		}
+	}
+	return nil
+}
+
+func init() {
+	mapValidateDescendants[MediaTypeManifest] = validateManifestDescendants
+}
```

Comment:
No need for `init()`, you can [declare a package-level constant][1].

[1]: https://github.com/wking/image-tools/blob/callback-validation/validate/validate.go#L30-L35